src/metabase/pulse/render/common.clj

```diff
@@ -22,11 +26,61 @@
   Object
   (toString [_] num-str))
 
+(defn number-formatter
+  "Return a function that will take a number and format it according to its column viz settings. Useful to compute the
+  format string once and then apply it over many values."
+  ([column col-viz]
+   (number-formatter column
+                     col-viz
+                     (mb.viz/db->norm-column-settings-entries
+                      (:type/Number (public-settings/custom-formatting)))))
```

Comment:
Yeah. I read over the xlsx stuff and I was worried that combining the two would make everything harder to read and not work as well. The xlsx stuff only returns format strings and then sets them on some reused cell objects and the requirements to unify it felt like everything would be harder to understand with lots of stuff passed through to keep track if we wanted an xlsx format string, a value formatting string, etc.

I think you might be right about the merge-global-settings though.

Reply:
Actually, I think there might be another way to approach this. I suspect we don't want to read this setting a bunch of times. It might be better to have the `process-viz-settings?` middleware to grab the settings once rather than repeatedly in other places. then a simple helper function can get the format settings for a column by getting by column id or name and merging the global settings.